### ISSUE TITLE
Improve homepage load speed

### ISSUE THREAD

**Issue by psiemens** [ubyssey/ubyssey-dispatch-theme#27]
The homepage is loading slowly in production and I suspect that it's due to a high number of complex database queries. It'd be great if we could find a way to simplify the queries being made in [this view](https://github.com/ubyssey/ubyssey-dispatch-theme/blob/staging/ubyssey/views.py#L43-L85), which includes a call to `ArticleHelper.get_frontpage`.

Keep in mind that I could be wrong and it might be something other than the queries. We just need to make the page fast again 😄

**Comment by mchernoff** [comment]
No problems on my end with the home page speed, even after I added hundreds of articles.

The queries seem fairly simple but I made a few alterations to get_frontpage that improved its speed by a few hundred microseconds. Let's hope it scales.
Also created a Csv file ('querytime') for tracking query runtimes and graphing them in excel for future testing.
Changes are in branch 'frontpage_speed'

**Comment by psiemens** [comment]
@mchernoff Awesome -- great work.

I'll try to test this in production today -- fingers crossed it makes a difference!

**Comment by psiemens** [comment]
I ran your code on our production site and logged the results (the querytime file is super useful, btw). Here are the times compared to my local machine:

**Local** (Macbook Air, OSX 10.10.5, 8GB RAM)

```
0.089000,0.072000,
0.082000,0.065000,
1.213000,0.073000,
0.076000,0.036000,
0.105000,0.050000,
0.064000,0.043000,
0.118000,0.046000,
0.080000,0.044000,
0.110000,0.045000,
```

**Production** (Ubuntu on EC2 t2.small, 2GB RAM)

```
0.093000,0.046000,
0.068000,0.044000,
0.089000,0.044000,
0.087000,0.043000,
0.092000,0.045000,
0.110000,0.044000,
0.082000,0.043000,
0.083000,0.043000,
```

As you can see, there isn't much of a difference. The homepage was still taking ~5 seconds to load, so the problem is still there. However, it looks like it might not be database-related.

**Comment by psiemens** [comment]
After a few hours spent tinkering last night, I was able to reduce the number of SQL queries from ~85 to ~50, which resulted in a load time reduction of about 75%.

As it turns out, due to the lazy loading behaviour of Django's QuerySet object, the raw query for the homepage algorithm was being evaluated **five times** (!!!) for each page load. I bypassed lazy loading and reduced it to a single query, which had a huge impact on the over page speed.

The lazy loading also made our benchmarking attempts futile because each query was being made at the template level rather than in the view. 😢 

I also optimized the section queries to pre-fetch all related data for each article (author, image) in a single query instead of running a separate query for each article, which drastically reduced the number of queries. 

I'm going to mark this as closed for now. Thanks for the help @mchernoff. I'd be happy to go over the specific changes if you're curious about how Django's ORM works -- I definitely learned a lot in the process.
